app/models/grda_warehouse/tasks/generate_client_roi_authorizations_task.rb

```diff
@@ -0,0 +1,100 @@
+###
+# Copyright 2016 - 2024 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+# Build ROI values from the client. We change ROI authorization to the canonical record in the future, instead of
+# generating it from the client. At that point, this task can be removed
+module GrdaWarehouse::Tasks
+  class GenerateClientRoiAuthorizationsTask
+    def self.perform(...)
+      new.perform(...)
+    end
+
+    # @param client_ids only rebuild
+    def perform(client_ids: nil)
+      with_lock do
+        scope = destination_client_scope
+        scope = scope.where(id: client_ids) unless client_ids.nil?
+        scope.find_in_batches do |batch|
+          values = []
+          batch.each do |client|
+            result = process_client(client)
+            values << result if result
+          end
+          GrdaWarehouse::ClientRoiAuthorization.import(
+            values,
+            on_duplicate_key_update: {
+              conflict_target: [:destination_client_id],
+              columns: values.first&.keys&.excluding(:destination_client_id),
+            },
+          )
+          # cleanup orphans
+          orphan_scope = GrdaWarehouse::ClientRoiAuthorization.
+            where.not(destination_client_id: destination_client_scope.select(:id))
+          orphan_scope = orphan_scope.where(destination_client_id: client_ids) if client_ids
+          orphan_scope.delete_all
+        end
+      end
+    end
+
+    protected
+
+    # we only consider building ROI authorizations for destination clients
+    def destination_client_scope
+      GrdaWarehouse::Hud::Client.destination
+    end
+
+    def process_client(destination_client)
+      status = roi_status(destination_client)
+
+      return nil unless status
+
+      {
+        status: status,
+        destination_client_id: destination_client.id,
+        coc_codes: roi_coc_codes(destination_client),
+        starts_at: destination_client.consent_form_signed_on,
+        expires_at: roi_expiry_date(destination_client),
+        # maybe add file, other fields
+      }
+    end
+
+    def roi_expiry_date(client)
+      duration = GrdaWarehouse::Hud::Client.release_duration
+      case GrdaWarehouse::Hud::Client.release_duration
+      when 'One Year', 'Two Years'
+        return nil unless client.consent_form_signed_on
```

Comment:
This is interesting because it means the client didn't sign the consent, but we determined it was valid, and therefore we are granting an indefinite consent.  This would probably arise if the client's consent was uploaded before the release duration was set or if the configuration wasn't correctly enforcing collection of a signature date.
I don't have a better solution, but it might be worth calling out here.   

What would you think about returning `Date.current + 1.years` or `Date.current + 2.years`?  

That might be misleading since it'll be extended every day...  Maybe we should count from the release form `created_at` date if we have one in this situation....